packages/transport/src/utils/receive.ts

```diff
@@ -3,29 +3,44 @@ import { Root } from 'protobufjs/light';
 import { decode as decodeProtobuf, createMessageFromType } from '@trezor/protobuf';
 import { TransportProtocol } from '@trezor/protocol';
 
-async function receiveRest(
+import { success } from './result';
+import { AbstractApi } from '../api/abstract';
+
+async function receiveRest<T extends () => ReturnType<AbstractApi['read']>>(
     result: Buffer,
-    receiver: () => Promise<Buffer>,
+    receiver: T,
     offset: number,
     expectedLength: number,
     chunkHeader: Buffer,
-): Promise<void> {
+) {
     if (offset >= expectedLength) {
         return;
     }
-    const data = await receiver();
+    const readResult = await receiver();
+    if (!readResult.success) {
+        return readResult;
```

Comment:
If I understand correctly, returning unsuccessful result here is irrelevant, because `receiveRest` is used as a procedure, not a function. Therefore, even if one `receiveRest` iteration fails (without throwing an error unlike before), `receive` will still return success with half-empty buffer.

Also, I'd like to convert the recursion into a cycle in the future, but no reason to do it right now.